Owin.Security.Providers/Foursquare/FoursquareAuthenticationHandler.cs

```diff
@@ -0,0 +1,238 @@
+	{
+		private const String AuthorizationEndpoint = "https://foursquare.com/oauth2/authenticate";
+		private const String TokenEndpoint = "https://foursquare.com/oauth2/access_token";
+		private const String GraphApiEndpoint = "https://api.foursquare.com/v2/users/self";
+		private const String XmlSchemaString = "http://www.w3.org/2001/XMLSchema#string";
+
+		private readonly ILogger _logger;
+		private readonly HttpClient _httpClient;
+
+		public FoursquareAuthenticationHandler(HttpClient httpClient, ILogger logger)
+		{
+			this._httpClient = httpClient;
+			this._logger = logger;
+		}
+
+		public override async Task<Boolean> InvokeAsync()
+		{
+			if ((String.IsNullOrEmpty(this.Options.CallbackPath) == false) && (this.Options.CallbackPath == this.Request.Path.ToString()))
+			{
+				return await this.InvokeReturnPathAsync();
+			}
+
+			return false;
+		}
+
+		protected override async Task<AuthenticationTicket> AuthenticateCoreAsync()
+		{
+			this._logger.WriteVerbose("AuthenticateCore");
+
+			AuthenticationProperties properties = null;
+
+			try
+			{
+				String code = null;
+				String state = null;
+
+				var query = this.Request.Query;
+				var values = query.GetValues("code");
+
+				if ((values != null) && (values.Count == 1))
+				{
+					code = values[0];
+				}
+
+				values = query.GetValues("state");
+
+				if ((values != null) && (values.Count == 1))
+				{
+					state = values[0];
+				}
+
+				properties = this.Options.StateDataFormat.Unprotect(state);
+
+				if (properties == null)
+				{
+					return null;
+				}
+
+				// OAuth2 10.12 CSRF
+				if (this.ValidateCorrelationId(properties, this._logger) == false)
+				{
+					return new AuthenticationTicket(null, properties);
+				}
+
+				var tokenRequestParameters = new List<KeyValuePair<String, String>>()
+				{
+					new KeyValuePair<String, String>("client_id", this.Options.ClientId),
+					new KeyValuePair<String, String>("client_secret", this.Options.ClientSecret),
+					new KeyValuePair<String, String>("grant_type", "authorization_code"),
+					new KeyValuePair<String, String>("redirect_uri", this.GenerateRedirectUri()),
+					new KeyValuePair<String, String>("code", code),
+				};
+
+				var requestContent = new FormUrlEncodedContent(tokenRequestParameters);
+
+				var response = await this._httpClient.PostAsync(TokenEndpoint, requestContent, this.Request.CallCancelled);
+				response.EnsureSuccessStatusCode();
+
+				var oauthTokenResponse = await response.Content.ReadAsStringAsync();
+
+				var oauth2Token = JObject.Parse(oauthTokenResponse);
+				var accessToken = oauth2Token["access_token"].Value<String>();
+
+				if (String.IsNullOrWhiteSpace(accessToken) == true)
+				{
+					this._logger.WriteWarning("Access token was not found");
+					return new AuthenticationTicket(null, properties);
+				}
+
+				var graphResponse = await this._httpClient.GetAsync(GraphApiEndpoint + "?oauth_token=" + Uri.EscapeDataString(accessToken), Request.CallCancelled);
```

Comment:
No, it doesn't, but I never ran it in localhost... I use 127.0.0.1.xip.io and configure IIS Express accordingly. Can you try it with a non-local address, please?